### PR TITLE
Addition of PERCENT keyword in the LIMIT clause

### PR DESCRIPTION
**Pull request by chenxi8611**
I've added percent keyword in the LIMIT clause.
@Mytherin -- please check it if it is implemented properly.

### CODE REVIEW

File: src/planner/binder/query_node/plan_subquery.cpp [hunk 1]
```diff
@@ -22,7 +22,7 @@ static unique_ptr<Expression> PlanUncorrelatedSubquery(Binder &binder, BoundSubq
 	case SubqueryType::EXISTS: {
 		// uncorrelated EXISTS
 		// we only care about existence, hence we push a LIMIT 1 operator
-		auto limit = make_unique<LogicalLimit>(1, 0, nullptr, nullptr);
+		auto limit = make_unique<LogicalLimit>(false, 1, 0, nullptr, nullptr);
```
Comment (Mytherin) [inline src/planner/binder/query_node/plan_subquery.cpp:25]: Also create a `<LogicalLimitPercentage>` and leave the regular Limit as-is. That way the distinction between percentage and non-percentage based limits is more obvious and not just a boolean flag.

File: test/sql/order/test_limit_percent.test [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+----
+11
+12
+13
+14
+15
+
+# LIMIT PERCENT with order by
+query I
+SELECT i from generate_series(1, 2000) tbl(i) WHERE (i / 10 * 10 = i) ORDER BY i DESC LIMIT 10%
```
Comment (Mytherin) [inline test/sql/order/test_limit_percent.test:45]: Please add the following tests:

* Tests with many values, e.g.

```sql
SELECT * FROM range(1000000) LIMIT 0.0001%
```

* Tests with LIMIT% and OFFSET combined
* LIMIT% with complex types (strings, lists, structs)
* LIMIT 0%
* Test behavior when percentage does not match exactly, e.g. if we have 10 rows, does 25% return 2 or 3 rows? What about 21%? Or 29%? That behavior should probably match Oracle/other systems that support this feature.
* LIMIT% in subqueries, e.g.:
```sql
SELECT * FROM (SELECT * FROM (SELECT * FROM range(1000000) LIMIT 10%) LIMIT 10%) LIMIT 10%;
```
* LIMIT% in set operations, e.g.:
```sql
SELECT * FROM range(10) UNION ALL SELECT * FROM range(10, 20) LIMIT 10%;
```

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR!

I see that you can have gone for the approach of materializing everything and then selecting the top X% of the materialized result. I think that is fine and a good initial approach.

I have added some comments.

Also note that the vector_sizes test is failing. Please have a look at that. You can test that locally by changing the vector_size to 2 in `src/include/duckdb/common/vector_size.hpp`, e.g.:

```
#define STANDARD_VECTOR_SIZE 2
```

File: test/sql/order/test_limit_percent.test (hunk 2 above)
Comment (chenxi8611) [reply test/sql/order/test_limit_percent.test:45]: Okay, I will add test case as you said.

File: src/execution/operator/helper/physical_limit.cpp [hunk 3]
```diff
@@ -109,6 +117,47 @@ SinkResultType PhysicalLimit::Sink(ExecutionContext &context, GlobalSinkState &g
 	return SinkResultType::NEED_MORE_INPUT;
 }
 
+void PhysicalLimit::Combine(ExecutionContext &context, GlobalSinkState &gstate, LocalSinkState &lstate) const {
+	auto &state = (LimitGlobalState &)gstate;
```
[the remark this answers is not in the document]
Comment (chenxi8611) [reply src/execution/operator/helper/physical_limit.cpp:121]: Thanks for your pointers. I am now considering move it to Finalize, because it is the cause of vector_sizes test failure.

**Comment by chenxi8611** [comment]
@Mytherin I've separately created the classes for limit percent, e.g. LimitPercentModifier, BoundLimitPercentModifier, LogicalLimitPercent, LogicalTopNPercent as you said. And I've added the test cases and fixed vector_size's error.
Please review it.

File: src/execution/operator/helper/physical_limit.cpp [hunk 4]
```diff
@@ -13,14 +13,32 @@ namespace duckdb {
 class LimitGlobalState : public GlobalSinkState {
 public:
 	explicit LimitGlobalState(const PhysicalLimit &op) : current_offset(0) {
-		this->limit = op.limit_expression ? INVALID_INDEX : op.limit_value;
-		this->offset = op.offset_expression ? INVALID_INDEX : op.offset_value;
+		if (!op.limit_expression) {
+			this->limit = op.limit_value;
```
Comment (Mytherin) [inline src/execution/operator/helper/physical_limit.cpp:17]: Please create a `PhysicalLimitPercent` that runs the limit with a percentage and revert any changes made to the physical limit. This switching on two entirely different modes of execution inside one operator is not ideal.

File: src/common/enums/logical_operator_type.cpp [hunk 5]
```diff
@@ -40,8 +40,12 @@ string LogicalOperatorToString(LogicalOperatorType type) {
 		return "ORDER_BY";
 	case LogicalOperatorType::LOGICAL_TOP_N:
 		return "TOP_N";
+	case LogicalOperatorType::LOGICAL_TOP_N_PERCENT:
+		return "TOP_N_PERCENT";
```
Comment (Mytherin) [inline src/common/enums/logical_operator_type.cpp:44]: TOP_N_PERCENT is not necessary for now.

File: src/execution/operator/order/physical_top_n.cpp [hunk 6]
```diff
@@ -260,6 +268,9 @@ void TopNHeap::Finalize() {
 }
 
 void TopNHeap::Reduce() {
+	if (limit == NumericLimits<idx_t>::Maximum()) {
```
Comment (Mytherin) [inline src/execution/operator/order/physical_top_n.cpp:271]: Please revert any changes made to the Top N operator. A Top-N operator that is optimized for a percentage could be created in a later PR but for now leave it as-is.

File: src/parallel/executor.cpp [hunk 7]
```diff
@@ -675,4 +677,168 @@ unique_ptr<DataChunk> Executor::FetchChunk() {
 	return chunk;
 }
 
+idx_t Executor::CountChunk() {
```
Comment (Mytherin) [inline src/parallel/executor.cpp:680]: Why are all these changes being made to the executor? This is out of scope for this PR and should not be necessary. Please revert.

File: src/planner/binder/query_node/bind_select_node.cpp [hunk 8]
```diff
@@ -48,6 +48,25 @@ unique_ptr<Expression> Binder::BindDelimiter(ClientContext &context, unique_ptr<
 	return expr;
 }
 
+unique_ptr<Expression> Binder::BindDelimiter(ClientContext &context, unique_ptr<ParsedExpression> delimiter,
+                                             double &delimiter_value) {
+	auto new_binder = Binder::CreateBinder(context, this, true);
+	ExpressionBinder expr_binder(*new_binder, context);
+	expr_binder.target_type = LogicalType::DOUBLE;
+	auto expr = expr_binder.Bind(delimiter);
+	if (expr->IsFoldable()) {
```
Comment (Mytherin) [inline src/planner/binder/query_node/bind_select_node.cpp:57]: This is exactly the same code as `Binder::BindDelimiter<int64_t>` but with a minor change that the LogicalType is different, could you turn this into a templated function instead to avoid unnecessary code duplication?

File: src/planner/binder/query_node/plan_select_node.cpp [hunk 9]
```diff
@@ -97,21 +98,33 @@ unique_ptr<LogicalOperator> Binder::CreatePlan(BoundSelectNode &statement) {
...
-			statement.modifiers.erase(statement.modifiers.end() - 1);
-			i--;
+		if (modifier->type == ResultModifierType::LIMIT_MODIFIER) {
+			auto &limit_modifier = (BoundLimitModifier &)*modifier;
+			if (limit_modifier.limit || limit_modifier.offset) {
+				PlanSubqueries(&limit_modifier.limit, &root);
+				PlanSubqueries(&limit_modifier.offset, &root);
+				auto limit = make_unique<LogicalLimit>(limit_modifier.limit_val, limit_modifier.offset_val,
+				                                       move(limit_modifier.limit), move(limit_modifier.offset));
+				limit->AddChild(move(root));
```
Comment (Mytherin) [inline src/planner/binder/query_node/plan_select_node.cpp:108]: Most of the code here is repeated. Could you extract the repeated code?

File: src/planner/binder/query_node/plan_select_node.cpp [hunk 10]
```diff
@@ -97,21 +98,33 @@ unique_ptr<LogicalOperator> Binder::CreatePlan(BoundSelectNode &statement) {
...
+				root = move(limit);
+				// Delete from modifiers
+				std::swap(statement.modifiers[i], statement.modifiers.back());
+				statement.modifiers.erase(statement.modifiers.end() - 1);
+				i--;
+			}
+		} else if (modifier->type == ResultModifierType::LIMIT_PERCENT_MODIFIER) {
+			auto &limit_modifier = (BoundLimitPercentModifier &)*modifier;
+			if (limit_modifier.limit) {
+				PlanSubqueries(&limit_modifier.limit, &root);
```
Comment (Mytherin) [inline src/planner/binder/query_node/plan_select_node.cpp:118]: PlanSubqueries on the `offset` also needs to happen here.

File: test/sql/order/test_limit_percent.test [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+SELECT * FROM test LIMIT RANDOM() %;
+
+# subquery in limit percent and offset
+query I
+SELECT a FROM test LIMIT 50% OFFSET 2
+----
+13
+
+query I
+SELECT a FROM test LIMIT (SELECT COUNT(*) FROM test) % OFFSET 1
```
Comment (Mytherin) [inline test/sql/order/test_limit_percent.test:116]: Could you add a test that has a subquery in the offset clause?

**Review by Mytherin — COMMENTED** [review]
Thanks for the changes! Some comments:

File: src/execution/operator/helper/physical_limit.cpp (hunk 4 above)
Comment (chenxi8611) [reply src/execution/operator/helper/physical_limit.cpp:17]: Okay, I will add PhysicalLimitPercent.

File: src/execution/operator/order/physical_top_n.cpp (hunk 6 above)
Comment (chenxi8611) [reply src/execution/operator/order/physical_top_n.cpp:271]: Okay. I will revert any changes made to the Top N operator.
By the way, I have a question.
Then can't use at same time "order by" and "limit %" clause. isn't it?

File: src/parallel/executor.cpp (hunk 7 above)
Comment (chenxi8611) [reply src/parallel/executor.cpp:680]: This is necessary, because we must know all source count as the result of SQL execution without "limit %" before the entire SQL execution. This code doesn't affect other parts, only affects limit percent operator.

File: test/sql/order/test_limit_percent.test (hunk 11 above)
Comment (chenxi8611) [reply test/sql/order/test_limit_percent.test:116]: Okay, I will add a test that has a subquery in the offset clause

File: src/planner/binder/query_node/plan_select_node.cpp (hunk 9 above)
Comment (chenxi8611) [reply src/planner/binder/query_node/plan_select_node.cpp:108]: Okay, I will extract the repeated code

File: src/planner/binder/query_node/plan_select_node.cpp (hunk 10 above)
Comment (chenxi8611) [reply src/planner/binder/query_node/plan_select_node.cpp:118]: Great, it is the part I missed. I will fix it. Thanks.

File: src/execution/operator/order/physical_top_n.cpp (hunk 6 above)
Comment (Mytherin) [reply src/execution/operator/order/physical_top_n.cpp:271]: No, the Top N node is just an optimization. It is not required for the system to function.

File: src/parallel/executor.cpp (hunk 7 above)
Comment (Mytherin) [reply src/parallel/executor.cpp:680]: This approach is very unmaintainable, since the entire query execution logic is now duplicated for the sole purpose of a single operator. It does not seem efficient either, since the entire query is now run multiple times.

The limit % operator is not special, it can be implemented without any changes to the executor.

Comment (chenxi8611) [reply src/parallel/executor.cpp:681]: Okay, I will find another way without using of executor

**Comment by chenxi8611** [comment]
I've removed PhysicalTopNPercent and added PhysicalLimitPercent, and I've fixed some errors and added a test that has a subquery in the offset clause. Also I've changed the code so that don't use executor.
@Mytherin -- Please review it.
